database/schema.sql

```diff
@@ -0,0 +1,181 @@
+    --
+    id    INTEGER,
+    desc  TEXT     null,
+
+    primary key (id asc)
+);
+
+-- Creating 'Subjects' table.
+--
+create table if not exists Subjects (
+    id         INTEGER,
+    course_id  INTEGER  not null,
+
+    primary key (id asc),
+    foreign key (course_id) references Courses (id)
+);
+
+create index if not exists Subject_course_ids on Subjects (course_id);
+
+-- Creating 'Students' table.
+--
+create table if not exists Students (
+    -- The "primary key" doesn't make field non null in sqlite.
+    --
+    addr  BLOB  not null,
+
+    primary key (addr)
+
+) without rowid;
+
+-- Creating 'StudentCourses' table.
+--
+-- [Appears to be] many-to-many ref between Students and Cources.
+--
+create table if not exists StudentCources (
+    student_addr  BLOB     not null,
+    course_id     INTEGER  not null,
+
+    -- Composite primary.
+    --
+    primary key (student_addr, course_id),
+
+    foreign key (student_addr) references Students (addr),
+    foreign key (course_id)    references Courses  (id)
+
+) without rowid;
+
+create index if not exists StudentCources_student_addr on StudentCources (student_addr);
+create index if not exists StudentCources_course_id    on StudentCources (course_id);
+
+-- Creating 'Assignments' table.
+--
+create table if not exists Assignments (
+    hash       BLOB     not null,
+    course_id  INTEGER  not null,
+    type       INTEGER  not null,
+    desc       TEXT     null,
+
+    primary key (hash),
+    foreign key (course_id) references Courses(id)
+
+) without rowid;
+
+create index if not exists Assigments_course_id on Assignments (course_id);
+
+-- Creating 'StudentAssignments' table.
+--
+-- [Appears to be] many-to-many ref between Students and Assigments.
+--
+create table if not exists StudentAssignments (
+    student_addr     BLOB  not null,
+    assignment_hash  BLOB  not null,
+
+    primary key (student_addr, assignment_hash),
+
+    foreign key (student_addr)    references Students    (addr)
+    foreign key (assignment_hash) references Assignments (hash)
+
+) without rowid;
+
+create index if not exists StudentAssigments_student_addr    on StudentAssignments (student_addr);
+create index if not exists StudentAssigments_assignment_hash on StudentAssignments (assignment_hash);
+
+-- Creating 'Submissions' table.
+--
+create table if not exists Submissions (
+    hash             BLOB     not null,
+    student_addr     BLOB     not null,
+    assignment_hash  BLOB     not null,
+    type             INTEGER  not null,
+    signature        BLOB     not null,
+
+    primary key (hash),
+    foreign key (student_addr)    references Students    (addr),
+    foreign key (assignment_hash) references Assignments (hash),
+
+    -- Composite foreign key.
+    -- We may not need it, actually, but it may help us.
+    --
+    foreign key (student_addr, assignment_hash)
+        references StudentAssignments(student_addr, assignment_hash)
```

Comment:
I think we should either use only this composite foreign key (and get student and assignment for submission through join with `StudentAssignments`, or use simple foreign keys. I think we might go the first way - it's more semantically valid (I just didn't think of it while I was doing PDM).